trace/trace_test.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package trace
+
+import (
+	"testing"
+	"time"
+)
+
+func TestNew(t *testing.T) {
+	tests := map[string]*struct {
+		inputString   string
+		expectedTrace *Trace
+	}{
+		"Test 1: When string is empty": {
+			inputString: "",
+			expectedTrace: &Trace{
+				name:  "",
+				steps: nil,
+			},
+		},
+		"Test 2: When string is not empty": {
+			inputString: "test string",
+			expectedTrace: &Trace{
+				name:  "test string",
+				steps: nil,
+			},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := New(tt.inputString)
+			if got.name != tt.expectedTrace.name {
+				t.Fatalf("Failed %s:\n Expected %v \n Got %v \n", name, tt.expectedTrace, got)
+			}
+		})
+	}
+}
+
+func TestStep(t *testing.T) {
+	sampleTrace := &Trace{}
+	tests := map[string]*struct {
+		inputString   string
+		expectedTrace *Trace
+	}{
+		"Test 1: When string is empty": {
+			inputString: "",
+			expectedTrace: &Trace{
+				steps: []traceStep{
+					{time.Now(), ""},
+				},
+			},
+		},
+		"Test 2: When string is not empty": {
+			inputString: "test2",
+			expectedTrace: &Trace{
+				steps: []traceStep{
+					{time.Now(), ""},
+					{time.Now(), "test2"},
+				},
+			},
+		},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			sampleTrace.Step(tt.inputString)
+
+			if sampleTrace.steps == nil || len(sampleTrace.steps) != len(tt.expectedTrace.steps) || sampleTrace.steps[len(sampleTrace.steps)-1].msg != tt.expectedTrace.steps[len(tt.expectedTrace.steps)-1].msg {
+				t.Fatalf("Failed %s:\n Expected %v \n Got %v \n", name, tt.expectedTrace, sampleTrace)
```

Comment:
This can be `t.Errorf`, generally `t.Fatal` would be used if you want to fail the test early (i.e. some setup method needed for the test case errors).

Reply:
cool